examples/ReinterpretLog.hs

```diff
@@ -0,0 +1,159 @@
+  Debug message -> "[debug] " ++ message
+  Info  message -> "[info] "  ++ message
+
+-- The application: it logs two messages, then quits.
+application :: (Carrier sig m, Member (Log Message) sig) => m ()
+application = do
+  log (Debug "debug message")
+  log (Info "info message")
+
+-- The application runner. Interpret the application by:
+--
+-- * Reinterpreting 'Log Message' effects as 'Log String' effects.
+-- * Interpreting 'Log String' effects by printing to stdout.
+runApplication :: IO ()
+runApplication =
+  application
+    & reinterpretLog renderLogMessage
+    & runLogStdout
+    & runM
+
+
+--------------------------------------------------------------------------------
+-- The logging effect
+--------------------------------------------------------------------------------
+
+-- Log an 'a', then continue with 'k'.
+data Log :: Type -> (Type -> Type) -> Type -> Type where
+  Log :: a -> k -> Log a m k
+  deriving stock (Functor)
+
+instance Effect (Log a) where
+  handle ::
+       Functor f
+    => f ()
+    -> (forall x. f (m x) -> n (f x))
+    -> Log a m (m b)
+    -> Log a n (n (f b))
+  handle state handler (Log x k) =
+    Log x (handler (k <$ state))
+
+instance HFunctor (Log a) where
+  hmap :: (forall x. m x -> n x) -> Log a m k -> Log a n k
+  hmap _ =
+    coerce
+
+-- Log an 'a'.
+log :: (Carrier sig m, Member (Log a) sig) => a -> m ()
+log x =
+  send (Log x (pure ()))
+
+
+--------------------------------------------------------------------------------
+-- The logging effect carriers
+--------------------------------------------------------------------------------
+
+-- Carrier one: log strings to stdout.
+newtype LogStdoutC m a
+  = LogStdoutC (m a)
+  deriving newtype (Applicative, Functor, Monad, MonadIO)
+
+runLogStdout :: LogStdoutC m a -> m a
+runLogStdout (LogStdoutC m) =
+  m
+
+instance
+     (Carrier sig m, MonadIO m)
+  => Carrier (Log String :+: sig) (LogStdoutC m) where
+
+  eff :: (Log String :+: sig) (LogStdoutC m) (LogStdoutC m a) -> LogStdoutC m a
+  eff = \case
+    L (Log message k) ->
+      LogStdoutC $ do
+        liftIO (putStrLn message)
+        runLogStdout k
+
+    R other ->
+      LogStdoutC (eff (handlePure runLogStdout other))
+
+
+-- Carrier two: reinterpret a program that logs 's's into one that logs 't's
+-- using a function from 's' to 't'.
+newtype ReinterpretLogC s t m a
+  = ReinterpretLogC ((s -> t) -> m a)
+  deriving stock (Functor)
+  deriving (Applicative, Monad, MonadIO)
+       via (ReaderT (s -> t) m)
+
+reinterpretLog :: (s -> t) -> ReinterpretLogC s t m a -> m a
+reinterpretLog f (ReinterpretLogC m) =
+  m f
+
+instance
+     (Carrier sig m, Member (Log t) sig)
+  => Carrier (Log s :+: sig) (ReinterpretLogC s t m) where
+
+  eff ::
+       (Log s :+: sig) (ReinterpretLogC s t m) (ReinterpretLogC s t m a)
+    -> ReinterpretLogC s t m a
+  eff = \case
+    L (Log s k) ->
+      ReinterpretLogC $ \f -> do
+        log (f s)
+        reinterpretLog f k
```

Comment:
If `ReinterpretLogC` is defined as a wrapper around `ReaderC`, its `Carrier` instance can be defined in terms of `Reader`, e.g. by using `ask` to retrieve the function.